File: Level3/그래프/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = 0
    # 노드 1부터 각 노드까지의 거리
    distance=[0]*(n+1)
    # 그래프 정보
    graph=[[] for _ in range(n+1)]
    # 방문 탐색할 queue
    queue=deque()
    
    for a,b in edge:
        # 양방향 그래프
        graph[a].append(b)
        graph[b].append(a)
    
    queue.append(1)
    distance[1]=1
    
    while queue:
        now=queue.popleft()
        for g in graph[now]:
            if distance[g]==0:
                queue.append(g)
                distance[g]=distance[now]+1
    
    # 가장 멀리떨어진 노드 거리
    max_edge=max(distance)
    # 개수
    for d in distance:
        if d==max_edge:
            answer+=1
        
    return answer

File: Level3/그래프/가장 먼 노드.js
class Queue {
  constructor() {
    this.storage = {};
    this.front = 0;
    this.rear = 0;
  }
  
  size() {
    if (this.storage[this.rear] === undefined) {
      return 0;
    } else {
      return this.rear - this.rear + 1;
    }
  }
  
  add(value) {
    if (this.size() === 0) {
      this.storage['0'] = value;
    } else {
      this.rear += 1;
      this.storage[this.rear] = value;
    }
  }
  
  popleft() {
    let temp;
    if (this.front === this.rear) {
      temp = this.storage[this.front];
      delete this.storage[this.front];
      this.front = 0;
      this.rear = 0;
    } else {
      temp = this.storage[this.front];
      delete this.storage[this.front];
      this.front += 1;
    }
    return temp;
  }
}

function BFS(start, checked, costList, graph){
    var queue=new Queue()
    var start=start
    var max=-1
    checked[start]=true
    queue.add([start,0])
    while (queue.size()>0){
        var tmp=queue.popleft()
        var node= tmp[0]
        var cost = tmp[1]
        costList[node]=cost
        var max=cost >max ? cost : max
        for (var i of graph[node]){
            if(checked[i]==false){
                checked[i]=true
                queue.add([i,cost+1])
            }
        }
    }
    return costList.filter((x)=>x==max).length
}

function solution(n, edge) {
    var graph=[]
    var checked=[]
    var costList=[]
    for (let i=0; i<n+1; i++){
        graph.push([])
        checked.push(false)
        costList.push(0)
    }
    for (let i=0; i<edge.length ; i++){
        var x=edge[i][0]
        var y=edge[i][1]
        graph[x].push(y)
        graph[y].push(x)
    }
   
    
    return BFS(1, checked, costList, graph);
}
